File: python/gtk4/oop/application.py
import gi, sys
gi.require_version("Gtk", "4.0")
from gi.repository import Gtk

class App(Gtk.Application):
    APP_ID = "io.github.Miqueas.GTK-Examples.Python.Gtk4.OOP.Application"

    def __init__(self):
        super().__init__(application_id = self.APP_ID)

    def do_activate(self):
        print("Gio.Application::activate called")

    def do_startup(self):
        # Ensures chain-up
        Gtk.Application.do_startup(self)
        print("Gio.Application::startup called")

if __name__ == "__main__":
    sys.exit(App().run(sys.argv))

File: cpp/gtk4/oop/application.cpp
#include <print>
#include <gtkmm.h>

static const Glib::ustring APP_ID = "io.github.Miqueas.GTK-Examples.C.Gtk4.OOP.Application";

class App : public Gtk::Application {
  public:
    static Glib::RefPtr<App> create() {
      return Glib::make_refptr_for_instance<App>(new App());
    }

  protected:
    App() : Gtk::Application(APP_ID) {}

    void on_startup() override {
      // Ensures chain-up
      Gtk::Application::on_startup();
      std::println("::startup");
    }

    void on_activate() override {
      std::println("::activate");
    }
};

int main(int argc, char** argv) {
  return App::create()->run(argc, argv);
}